### ISSUE TITLE
dynamic sensors/rules

### ISSUE THREAD

**Issue by AndyMoore** [StackStorm/st2#4389]
##### SUMMARY

Trying to find a way to implement dynamic sensors/rules

##### ISSUE TYPE

The kubernetes pack has a sensor that watches for the creation of third party resources (TPR - a type of resource within kubernetes that makes the api extensible). When a TPR is created, a new API endpoint is created within kubernetes, and the stackstorm kubernetes pack creates a new sensor to watch that endpoint.

When an instance of a TPR is created, we have a second pack thats specific to our platform that watches these new sensors via rules/triggers, and reacts to additions to them.

The problem lies in the ordering. When a trigger is fired, a rule should spot it. That rule needs to exist before the trigger is fired, otherwise it misses the event, and a sensor restart is required (after the rule has been created) to re-emit the trigger automatically. 

So the question becomes how can I create a rule that kicks off a workflow based on a generated sensor, when the sensor doesn't exist when stackstorm is started?

**Comment by nmaludy** [comment]
Thoughts:
- Create and register the Trigger
- Create and register the Rule
- Create, register and start the Sensor


This will have the trigger (root dependency) avail first. Then the rule will be created so that no triggers are missed for processing. Finally the sensor will be created to emit the triggers.

**Comment by AndyMoore** [comment]
the problem with that is that the rule and sensor are in 2 separate packs. There's no dependancy management between packs, or resources within packs.

Further, the trigger/sensor are dynamic, so don't exist when stackstorm starts. The rules, however, would exist (because we know what resources are in our kubernetes deployment, and therefore which sensors will be created and what the trigger names will be). 

The rules won't be added because the triggers don't exist (and in fact st2ctl reload bombs out at this point)

The only real solution as far as I can tell is to know the triggers in advance, but the idea of a dynamic pack is that I can have resources called apple, banana and orange, and you can have some called car, plane and train, and they'd all have sensors created for them

**Comment by nmaludy** [comment]
Can you provide a minimum example to reproduce this issue?

**Comment by AndyMoore** [comment]
create a rule for a trigger that doesnt exist, and then create the trigger

**Comment by AndyMoore** [comment]
the other option is if i can find a reliable way to tell when a sensor has been created, and create dynamic rules within our own pack

maybe something like core.st2.sensor.process_spawn

**Comment by nmaludy** [comment]
I'm still not clear why the trigger can't be created before the rule?

**Comment by AndyMoore** [comment]
the trigger would need to be in the kubernetes pack. the rule is in the bitesize pack. you can't create a kubernetes trigger in the bitesize pack, and the kubernetes pack doesnt know what triggers it'll dynamically create when it starts up (and new ones can be created over time)

**Comment by nmaludy** [comment]
So, you're asking for rules to allow "late binding" and not validate trigger types when registered?

**Comment by AndyMoore** [comment]
yes. i want to be able to create a rule for a trigger/sensor i know will be created in the future
